### PR TITLE
[extractor/youtube] Improve channel tab handling

### PR DESCRIPTION
**Pull request by coletdjnz**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

#### Important/Main changes:
- If a user provides a url to a channel without a tab, yt-dlp will also try to download the streams and shorts tabs along with videos tab.
    - E.g. passing https://www.youtube.com/channel/UCK9V2B22uJYu3N7eR_BT9QA will download the videos, shorts and live tabs.
- The extractor will no longer try to fall back on the home page if none of the videos, shorts and streams tabs and uploads playlist exists. 
   -  This ensures you'll always get all the channels uploads by default.
      - One exception: special destination "channels" (see below) that have an invisible tab
   -  **IMPORTANT**: Channels with no videos **will raise an error** as none of the above will exist. Previously youtube would provide an empty videos tab (yt-dlp would get 0 videos), but that is no longer the case.

There is now more robust handling so the tab IE can handle even more weird channels such as some of the destination ones:
- https://www.youtube.com/channel/UCQvWX73GQygcwXOTSf_VDVg/letsplay
- https://www.youtube.com/channel/UCtFRv9O2AHqOZjjynzrv-xg
- https://www.youtube.com/channel/UCrpQ4p1Ql_hG8rKXIKM1MOQ - will now fail unless /featured tab is given

Fixes https://github.com/yt-dlp/yt-dlp/issues/5419, Fixes https://github.com/yt-dlp/yt-dlp/issues/5428, Fixes https://github.com/yt-dlp/yt-dlp/issues/5430

Supersedes https://github.com/yt-dlp/yt-dlp/pull/5432

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by coletdjnz** [comment]
TODO: 
- probably need to edit the README regarding the channel redirect compat opt, in case we changed anything it says
- @pukkandan yt-dlp cannot handle the case where channels have videos and streams and/or shorts tabs. Something to do with semi-resolved playlist within a playlist.

**Comment by Lesmiscore** [comment]
I think the people at NewPipe suggested to use the corresponding UU playlists
Can't it be used instead if the user hit the "bad" course?

**Comment by coletdjnz** [comment]
> I think the people at NewPipe suggested to use the corresponding UU playlists Can't it be used instead if the user hit the "bad" course?

Not quite sure what you mean :confused: 

We are using the UU/uploads playlist when neither of the 3 videos,streams and shorts tabs exist atm

File: yt_dlp/extractor/youtube.py [hunk 1]
```diff
@@ -5765,77 +5897,94 @@ def get_mobj(url):
...
         # YouTube sometimes provides a button to reload playlist with unavailable videos.
         if 'no-youtube-unavailable-videos' not in compat_opts:
             data = self._reload_with_unavailable_videos(item_id, data, ytcfg) or data
         self._extract_and_report_alerts(data, only_once=True)
-        tabs = traverse_obj(data, ('contents', 'twoColumnBrowseResultsRenderer', 'tabs'), expected_type=list)
+
+        tabs = self._extract_tab_renderers(data)
         if tabs:
-            return self._extract_from_tabs(item_id, ytcfg, data, tabs)
+            tab_results[:0] = [self._extract_from_tabs(item_id, ytcfg, data, tabs)]
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5977]: pls explain

**Comment by pukkandan** [comment]
We arent handling <https://www.youtube.com/gaming> correctly

**Comment by coletdjnz** [comment]
> We arent handling <https://www.youtube.com/gaming> correctly

tbf it's buggy on YouTube's end

**Comment by pukkandan** [comment]
Is that the only known page with this issue?

File: yt_dlp/extractor/youtube.py [hunk 2]
```diff
@@ -5765,77 +5897,94 @@ def get_mobj(url):
...
+                if requested_tab_id == 'live':
+                    # Live tab should have redirected to the video
+                    # Except in the case the channel has an actual live tab
+                    # Example: https://www.youtube.com/channel/UCEH7P7kyJIkS_gJf93VYbmg/live
+                    raise UserNotLive(video_id=mobj['id'])
+                elif selected_tab_name:
+                    raise ExtractorError(f'This channel does not have a {requested_tab_id} tab', expected=True)
+
+                # For channels such as https://www.youtube.com/channel/UCtFRv9O2AHqOZjjynzrv-xg
+                url = ''.join((pre, f'/{selected_tab_id}' if selected_tab_id else '', post))
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:5966]: pls explain

**Comment by coletdjnz** [comment]
> Is that the only known page with this issue?

as far as I know, yes

File: yt_dlp/extractor/youtube.py (hunk 2 above)
Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:5966]: this is for the case where we accept the home/default page. The tab url here would be /videos but we are downloading a different tab so we need to fix it.

Though the behaviour here rewrites it as https://www.youtube.com/channel/UCtFRv9O2AHqOZjjynzrv-xg/UCtFRv9O2AHqOZjjynzrv-xg. Whether that is a valid tab url on youtube's side is unknown, but yt-dlp will treat that as a direct tab url.

**Comment by pukkandan** [comment]
In that case, let it be ig

File: yt_dlp/extractor/youtube.py [hunk 3]
```diff
@@ -5705,6 +5706,120 @@ class YoutubeTabIE(YoutubeTabBaseInfoExtractor):
...
+            'channel': '中村悠一',
+            'uploader_url': 'https://www.youtube.com/channel/UC3eYAvjCVwNHgkaGbXX3sig',
+            'channel_follower_count': int,
+            'uploader': '中村悠一',
+            'description': 'md5:e744f6c93dafa7a03c0c6deecb157300',
+        },
+        'playlist_mincount': 60,
+    }, {
+        # Channel with no uploads and hence no videos, streams, shorts tabs or uploads playlist. This should fail.
+        # FIXME: how to test?
```
[the remark this answers is not in the document]
Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py]: might as well add a test to that. But I'll leave these in here for visibility

File: yt_dlp/extractor/youtube.py (hunk 2 above)
Comment (pukkandan) [reply yt_dlp/extractor/youtube.py:5968]: why not just strip the `/videos`?
```
url = f'{pre}{post}'
```

Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:5968]: was just trying to be general. that works too, but by yt-dlp that isn't really the correct url

**Comment by tzarebczan** [comment]
@coletdjnz - to be clear, ```@pukkandan yt-dlp cannot handle the case where channels have videos and streams and/or shorts tabs. Something to do with semi-resolved playlist within a playlist.``` is still not handled right? 

Here's an example: ```FOUND: [lbry@ytsync2:~] 1 $ yt-dlp --skip-download https://www.youtube.com/channel/UCAl9Ld79qaZxp9JzEOwd3aA/streams --get-id --flat-playlist --cookies cookies.txt --playlist-end 400 | grep "QfSbA_gQwW0"
QfSbA_gQwW0```

```NOT FOUND [lbry@ytsync2:~] 3s $ yt-dlp --skip-download https://www.youtube.com/channel/UCAl9Ld79qaZxp9JzEOwd3aA --get-id --flat-playlist --cookies cookies.txt --playlist-end 400 | grep "QfSbA_gQwW0"```

**Comment by pukkandan** [comment]
@tzarebczan Works fine. Sure you are on master branch? The change isnt released yet

```
❯ yt-dlp --skip-download https://www.youtube.com/channel/UCAl9Ld79qaZxp9JzEOwd3aA/streams --get-id --flat-playlist --playlist-end 10
QfSbA_gQwW0
LDtZyMcaaNY
yIxWvO3cgNg
Ytaa_5liwMA
CFdSIW8gQc0
PmWeXVRPeic
fxGOIJM5eF0
OhpEt6u8QAg
QkZUlqetTRA
9jk-O2zaAEQ
```

**Comment by tzarebczan** [comment]
yea, it comes up on /streams, but I was hoping it would also come up on / (without any tab). We'd rather make a single call for all uploads, vs trying to call each tab. I thought that was the purpose of this PR: ```if a user provides a url to a channel without a tab, yt-dlp will also try to download the streams and shorts tabs along with videos tab.
E.g. passing https://www.youtube.com/channel/UCK9V2B22uJYu3N7eR_BT9QA will download the videos, shorts and live tabs.```

**Comment by pukkandan** [comment]
hm.. The issue is `--flat-playlist`. I'll see what can be done

**Comment by pukkandan** [comment]
There are a few ways in which we can return the result of a channel URL; each with it's own issues. These are possible implementations I can think of

1. Return all entries in `/videos` and `url_result` of other tabs (**current implementation**)
    * `--flat-playlist --get-id` doesn't work like before, making it difficult to get all video ids
    * Returns playlists, possibly breaking old scripts that (incorrectly) didn't account for nested playlists
2. Squash all entries together and return as single playlist
    * reversed `playlist_index` changes when video is added to later tabs, make it hard to 
    * Data about the tab is lost (non-issue?)
3. Return all entries in all tabs irrespective of --flat-playlist ([POC](https://github.com/pukkandan/yt-dlp-dev/commit/6f2412d0305b6821d7f24d1246f56db75c9d6d44))
    * Unnecessary additional requests for -I 0 ([hacky workaround](https://github.com/pukkandan/yt-dlp-dev/commit/39a3f4a6110992e1d258728fe68d793b0f07b45e))
    * Returns playlists, possibly breaking old scripts that (incorrectly) didn't account for nested playlists
4. Return `UU` playlist (easiest solution tbh)
    * Breaks some metadata (there is a possible workaround)
    * Number of entries is limited (deal breaker?)

I'm leaning towards (3), but would like feedback

**Comment by krichbanana** [comment]
What I like about 4 is that newest videos always come first (from what I can tell), which is nice to have.

**Comment by pukkandan** [comment]
True. The 20k limit also won't matter for most users, but implementing that forces the limit on everyone. With other methods however, the user can always pass the UU playlist to yt-dlp. The only issue is that many people don't know the existence of this playlist

**Comment by emansom** [comment]
Option 2 has my vote, keeps the same behavior as old. Leads to minimal breakage?

**Comment by tzarebczan** [comment]
A sorted newest to oldest option would be the best for us also. It can be a new endpoint/param to look across all tab.

**Comment by coletdjnz** [comment]
> A sorted newest to oldest option would be the best for us also. It can be a new endpoint/param to look across all tab.

Unfortunately I don't think this is feasible or possible to add reliably. I recommend using the channel uploads playlist instead if it is an issue, or do custom processing of the results on your end.

**Comment by pukkandan** [comment]
My question with that is, why not use the UU playlist URL then?

**Comment by pukkandan** [comment]
I tried implementing each of the above, and here are my conclusions:

> 1. Return all entries in `/videos` and `url_result` of other tabs (**current implementation**)

This is strictly inferior to (3), and breaks a very common use-case. Not worth considering

> 2. Squash all entries together and return as single playlist

The only advantage this gives over (3) is the lack of nested playlists. But since the playlist returned is not sorted in any sensible way, this doesn't make much sense imho

> 4. Return `UU` playlist (easiest solution tbh)

This was supposed to have been the easiest method. But properly handling topic pages (which don't have a UU), and adding the workaround for metadata made it even more complex than (3). 

The biggest benefit of this is that it gives a single **sorted** playlist. But even when this is not implemented, users can always use the UU directly if a sorted playlist is desired. On the other hand, implementing this makes it so that there is no way to get all uploads of a channel with > 20k videos

So (3) does appear to be the solution that covers most use-cases

> The only issue is that many people don't know the existence of this playlist

#5487 prints a message letting users know of this

**Comment by emansom** [comment]
Is there a way to have the results from these playlists sorted by publication date?

**Pull request merged**